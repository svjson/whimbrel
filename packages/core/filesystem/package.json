{
  "name": "@whimbrel/filesystem",
  "version": "0.0.2",
  "description": "File System abstractions for Whimbrel",
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "test": "vitest run",
    "typecheck": "tsc --noEmit",
    "clean": "rimraf ./dist",
    "publish": "npm publish --access public"
  },
  "peerDependencies": {
    "@whimbrel/core-api": "^0.0.2"
  },
  "dependencies": {
    "@whimbrel/core-api": "*",
    "jsonc-parser": "^3.3.1"
  },
  "devDependencies": {
    "@whimbrel-test/tree-fixtures": "*",
    "@types/node": "^24.3.0",
    "vite": "^7.1.2",
    "vite-tsconfig-paths": "^5.1.4",
    "vitest": "^3.2.4"
  }
}
