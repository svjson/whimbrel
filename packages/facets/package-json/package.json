{
  "name": "@whimbrel/package-json",
  "version": "0.0.1",
  "description": "package.json Facet for Whimbrel",
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "test": "vitest run",
    "typecheck": "tsc --noEmit",
    "clean": "rimraf ./dist"
  },
  "peerDependencies": {
    "@whimbrel/core-api": "^0.0.1",
    "@whimbrel/facet": "^0.0.1",
    "@whimbrel/flow": "^0.0.1"
  },
  "dependencies": {
    "@whimbrel/core-api": "*",
    "@whimbrel/facet": "*",
    "@whimbrel/flow": "*"
  },
  "devDependencies": {
    "@whimbrel-test/tree-fixtures": "*",
    "@whimbrel-test/context-fixtures": "*",
    "@types/node": "^24.3.0",
    "vite": "^7.1.2",
    "vite-plugin-dts": "^4.5.4",
    "vite-tsconfig-paths": "^5.1.4",
    "vitest": "^3.2.4"
  }
}
