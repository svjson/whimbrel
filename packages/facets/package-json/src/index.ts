import { makeFacetModule, moduleTasks } from '@whimbrel/core-api'
import { detect, applyLicenseAugmentation } from './features'
import { AddScript, RemoveScript, SetProperty, SetVersion } from './tasks'
import { queryIndex } from './query'

export {
  isVersion,
  parseVersion,
  updateVersion,
  updateVersionString,
  versionString,
} from './lib'

export { PackageJSON, WorkspaceAdapter } from './adapters'

export { AddScript, RemoveScript, SetProperty, SetVersion }

export default makeFacetModule({
  id: 'package.json',
  implicits: [{ facet: 'node', scope: { roles: ['engine'] } }, 'project'],
  tasks: moduleTasks(AddScript, RemoveScript, SetProperty, SetVersion),
  detect,
  taskAugmentations: {
    'license:apply': {
      steps: applyLicenseAugmentation,
    },
  },
  queryIndex,
})
