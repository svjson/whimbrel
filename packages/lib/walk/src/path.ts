import type { PropertyPath } from './types'
import { walkPath } from './walk-path'

/**
 * Utility function for writing to a nested path on an object.
 * If the path does not exist, it will be created.
 *
 * ```ts
 * writePath({ nested: { things: {} }, 'nested.objects', 'here')
 * // => { nested: { things: {}, objects: 'here' }}
 * ```
 *
 * @param object - The object to write to.
 * @param propertyPath - The path to write to, either as a string or an array of segments.
 * @param value - The value to write at the specified path.
 *
 * @return The modified object with the value written at the specified path.
 */
export const writePath = (object: any, propertyPath: PropertyPath, value: any) => {
  return walkPath(
    object,
    propertyPath,
    (node, propName) => {
      node[propName] = {}
    },
    (node, propName) => (node[propName] = value)
  )
}

/**
 * Utility function for reading a value from a nested path on an object.
 * If the path does not exist, it will return null.
 * ```ts
 * readPath({ nested: { things: { here: 'value' } } }, 'nested.things.here')
 * // => 'value'
 *
 * readPath({ nested: { things: {} } }, 'nested.things.here')
 * // => null
 * ```
 *
 * @param object - The object to read from.
 * @param propertyPath - The path to read from, either as a string or an array of segments.
 *
 * @return The value at the specified path, or null if the path does not exist.
 */
export const readPath = (object: any, propertyPath: PropertyPath) => {
  return walkPath(object, propertyPath, null, (_parent, _propName, value) => value)
}

/**
 * Utility function for reading the closest value from a nested path on an object.
 * It will traverse the path and return the value of the last segment.
 * If the path does not exist, it will return null.
 *
 * ```ts
 * readClosest({ nested: { things: { here: 'value' } } }, 'nested.things.here')
 * // => 'value'
 *
 * readClosest({ nested: { things: {} } }, 'nested.things.here')
 * // => {}
 * ```
 *
 * @param object - The object to read from.
 * @param propertyPath - The path to read from, either as a string or an array of segments.
 *
 * @return The value at the closest segment of the specified path, or null if the path does not exist.
 */
export const readClosest = (object: any, propertyPath: PropertyPath) => {
  return walkPath(
    object,
    propertyPath,
    (node) => node,
    (_p, _pn, val) => val
  )
}

/**
 * Utility function for deleting a property at a nested path on an object.
 * If the path does not exist, it will do nothing.
 *
 * ```ts
 * deletePath({ nested: { things: { here: 'value' } } }, 'nested.things.here')
 * // => { nested: { things: {} } }
 *
 * deletePath({ nested: { things: {} } }, 'nested.things.here')
 * // => { nested: { things: {} } }
 * ```
 *
 * @param object - The object to delete from.
 * @param propertyPath - The path to delete from, either as a string or an array of segments.
 *
 * @return The modified object with the property deleted at the specified path.
 */
export const deletePath = (object: any, propertyPath: PropertyPath) => {
  return walkPath(object, propertyPath, null, (node, propName) => delete node[propName])
}
