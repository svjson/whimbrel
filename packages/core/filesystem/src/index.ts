import path from 'node:path'
import { FileSystem } from '@whimbrel/core-api'

export { ContextFileSystem } from './context'
export { DiskFileSystem, PhysicalFileSystem } from './physical'
export { MemoryFileSystem } from './memfs'
export { FileSystemReporter, FileSystemMutationReporter } from './reporter'

export const ifFileExistsAt = async <T>(
  disk: FileSystem,
  filePath: string | string[],
  fileName: string,
  fn: (fPath: string) => Promise<T>
) => {
  let actualPath = Array.isArray(filePath) ? path.join(...filePath) : filePath
  if (path.basename(actualPath) !== fileName) {
    actualPath = path.join(actualPath, fileName)
  }

  if (await disk.exists(actualPath)) {
    return fn(actualPath)
  }
}
