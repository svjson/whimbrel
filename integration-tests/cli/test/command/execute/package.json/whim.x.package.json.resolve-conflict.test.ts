import { describe, expect, test } from 'vitest'
import path from 'node:path'

import makeTreeFixture from '@whimbrel-test/tree-fixtures'
import { DiskFileSystem } from '@whimbrel/filesystem'
import { whimCli } from 'test/whim-fixture'
import { PackageJSON } from '@whimbrel/package-json'

const { createDirectory } = makeTreeFixture(DiskFileSystem)

describe('whim x package.json:resolve-conflict', () => {
  test('whim x package.json:resolve-conflict', async () => {
    // Given
    const root = await createDirectory(
      [{ 'package.base.json': '@facets/package.json/package.simple.json' }],
      DiskFileSystem
    )
    const oursPath = path.join(root, 'package.ours.json')
    const theirsPath = path.join(root, 'package.theirs.json')

    const basePath = path.join(root, 'package.base.json')

    const ours = await PackageJSON.read(DiskFileSystem, basePath)
    ours.set('version', '2.4.1')
    ours.set('dependencies.@koa/cors', '^5.1.0')
    ours.delete('dependencies.koa-bodyparser')
    await ours.write(oursPath)

    const theirs = await PackageJSON.read(DiskFileSystem, basePath)
    theirs.set('version', '2.4.0')
    theirs.set('dependencies.koa', '^2.16.4')
    theirs.set('devDependencies.ts-node', '^10.10.0')
    await theirs.write(theirsPath)

    const whim = whimCli(root)

    await whim.execute([
      'x',
      'package.json:resolve-conflict',
      '--base',
      'package.base.json',
      '--ours',
      'package.ours.json',
      '--theirs',
      'package.theirs.json',
    ])

    expect(whim.stdout).toMatchSnapshot()
    expect(await DiskFileSystem.readJson(oursPath)).toEqual({
      name: 'simple-things',
      version: '2.4.1',
      author: 'Benny Boxare',
      scripts: {
        build: 'tsc',
        dev: 'nodemon -e ts,js --exec ts-node -r dotenv/config src/index',
        test: 'vitest run',
      },
      dependencies: {
        '@koa/cors': '^5.1.0',
        '@koa/router': '^12.0.1',
        koa: '^2.16.4',
      },
      devDependencies: {
        '@types/koa': '^2.15.0',
        '@types/koa__cors': '^5.0.0',
        '@types/koa__router': '12.0.4',
        '@types/koa-bodyparser': '^4.3.12',
        '@types/node': '^20.11.5',
        'ts-node': '^10.10.0',
        typescript: '^5.6.2',
      },
    })
  })
})
