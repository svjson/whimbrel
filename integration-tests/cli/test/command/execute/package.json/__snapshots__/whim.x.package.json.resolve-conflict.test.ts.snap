// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`whim x package.json:resolve-conflict > whim x package.json:resolve-conflict 1`] = `
"
Execute Task: [32mpackage.json:resolve-conflict[39m - [34m/tmp/whim-dir-XXXX[39m
------------------------------------------------------------------
* Resolve package.json merge conlict
  ðŸ›¢ [34mM[39m /tmp/whim-dir-XXXX/package.ours.json

"
`;
